{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Notebook_6b_GenoML_Optional_Homework ",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/RafaelVillasmil/BIOF309_Introduction_to_Python/blob/main/Notebook_6b_GenoML_Optional_Homework.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JpNNM9-2p2FZ"
      },
      "source": [
        "## EXTRA POINTS (5): This is an example of a Machine Learning Pipeline using GenoML. Try to run it on your own"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "AqGlXavlbroA",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7373e945-404e-42f2-c20f-14b38e3d7e7f"
      },
      "source": [
        "import os\n",
        "import sys\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive/', force_remount=True)\n",
        "os.chdir(\"/content/drive/MyDrive/Colab_Notebooks/example_data/\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive/\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-KAJLMB7ckec"
      },
      "source": [
        "! mkdir outputs"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "UhupzPwabscP",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bc43f212-5cf8-48b8-8957-e3f07120f010"
      },
      "source": [
        "! pip install genoml2\n",
        "! pip install tables --upgrade\n",
        "! pip install umap-learn"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting genoml2\n",
            "  Downloading genoml2-1.0.0b11-py3-none-any.whl (72 kB)\n",
            "\u001b[K     |████████████████████████████████| 72 kB 1.2 MB/s \n",
            "\u001b[?25hCollecting pandas-plink\n",
            "  Downloading pandas_plink-2.2.9-cp37-cp37m-manylinux2010_x86_64.whl (100 kB)\n",
            "\u001b[K     |████████████████████████████████| 100 kB 10.2 MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from genoml2) (1.21.5)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.7/dist-packages (from genoml2) (1.1.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from genoml2) (2.23.0)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.7/dist-packages (from genoml2) (1.4.1)\n",
            "Requirement already satisfied: seaborn in /usr/local/lib/python3.7/dist-packages (from genoml2) (0.11.2)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from genoml2) (1.3.5)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.7/dist-packages (from genoml2) (3.2.2)\n",
            "Requirement already satisfied: tables in /usr/local/lib/python3.7/dist-packages (from genoml2) (3.7.0)\n",
            "Requirement already satisfied: statsmodels in /usr/local/lib/python3.7/dist-packages (from genoml2) (0.10.2)\n",
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.7/dist-packages (from genoml2) (1.0.2)\n",
            "Requirement already satisfied: xgboost in /usr/local/lib/python3.7/dist-packages (from genoml2) (0.90)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->genoml2) (2.8.2)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib->genoml2) (0.11.0)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->genoml2) (3.0.7)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->genoml2) (1.3.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.1->matplotlib->genoml2) (1.15.0)\n",
            "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.7/dist-packages (from pandas->genoml2) (2018.9)\n",
            "Requirement already satisfied: tqdm>=4.36.1 in /usr/local/lib/python3.7/dist-packages (from pandas-plink->genoml2) (4.63.0)\n",
            "Collecting zstandard>=0.13.0\n",
            "  Downloading zstandard-0.17.0-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.5 MB)\n",
            "\u001b[K     |████████████████████████████████| 2.5 MB 55.2 MB/s \n",
            "\u001b[?25hCollecting Deprecated>=1.2.6\n",
            "  Downloading Deprecated-1.2.13-py2.py3-none-any.whl (9.6 kB)\n",
            "Collecting pytest>=5.2.2\n",
            "  Downloading pytest-7.1.0-py3-none-any.whl (297 kB)\n",
            "\u001b[K     |████████████████████████████████| 297 kB 46.8 MB/s \n",
            "\u001b[?25hRequirement already satisfied: xarray>=0.18.2 in /usr/local/lib/python3.7/dist-packages (from pandas-plink->genoml2) (0.18.2)\n",
            "Requirement already satisfied: dask[array,dataframe]>=2.6.0 in /usr/local/lib/python3.7/dist-packages (from pandas-plink->genoml2) (2.12.0)\n",
            "Requirement already satisfied: cffi>=1.14.3 in /usr/local/lib/python3.7/dist-packages (from pandas-plink->genoml2) (1.15.0)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.7/dist-packages (from cffi>=1.14.3->pandas-plink->genoml2) (2.21)\n",
            "Collecting fsspec>=0.6.0\n",
            "  Downloading fsspec-2022.2.0-py3-none-any.whl (134 kB)\n",
            "\u001b[K     |████████████████████████████████| 134 kB 48.5 MB/s \n",
            "\u001b[?25hCollecting partd>=0.3.10\n",
            "  Downloading partd-1.2.0-py3-none-any.whl (19 kB)\n",
            "Requirement already satisfied: toolz>=0.7.3 in /usr/local/lib/python3.7/dist-packages (from dask[array,dataframe]>=2.6.0->pandas-plink->genoml2) (0.11.2)\n",
            "Requirement already satisfied: wrapt<2,>=1.10 in /usr/local/lib/python3.7/dist-packages (from Deprecated>=1.2.6->pandas-plink->genoml2) (1.13.3)\n",
            "Collecting locket\n",
            "  Downloading locket-0.2.1-py2.py3-none-any.whl (4.1 kB)\n",
            "Requirement already satisfied: tomli>=1.0.0 in /usr/local/lib/python3.7/dist-packages (from pytest>=5.2.2->pandas-plink->genoml2) (2.0.1)\n",
            "Requirement already satisfied: py>=1.8.2 in /usr/local/lib/python3.7/dist-packages (from pytest>=5.2.2->pandas-plink->genoml2) (1.11.0)\n",
            "Requirement already satisfied: importlib-metadata>=0.12 in /usr/local/lib/python3.7/dist-packages (from pytest>=5.2.2->pandas-plink->genoml2) (4.11.2)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.7/dist-packages (from pytest>=5.2.2->pandas-plink->genoml2) (21.3)\n",
            "Requirement already satisfied: attrs>=19.2.0 in /usr/local/lib/python3.7/dist-packages (from pytest>=5.2.2->pandas-plink->genoml2) (21.4.0)\n",
            "Collecting pluggy<2.0,>=0.12\n",
            "  Downloading pluggy-1.0.0-py2.py3-none-any.whl (13 kB)\n",
            "Requirement already satisfied: iniconfig in /usr/local/lib/python3.7/dist-packages (from pytest>=5.2.2->pandas-plink->genoml2) (1.1.1)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata>=0.12->pytest>=5.2.2->pandas-plink->genoml2) (3.7.0)\n",
            "Requirement already satisfied: typing-extensions>=3.6.4 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata>=0.12->pytest>=5.2.2->pandas-plink->genoml2) (3.10.0.2)\n",
            "Requirement already satisfied: setuptools>=40.4 in /usr/local/lib/python3.7/dist-packages (from xarray>=0.18.2->pandas-plink->genoml2) (57.4.0)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->genoml2) (2.10)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->genoml2) (1.24.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->genoml2) (2021.10.8)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->genoml2) (3.0.4)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn->genoml2) (3.1.0)\n",
            "Requirement already satisfied: patsy>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from statsmodels->genoml2) (0.5.2)\n",
            "Requirement already satisfied: numexpr>=2.6.2 in /usr/local/lib/python3.7/dist-packages (from tables->genoml2) (2.8.1)\n",
            "Installing collected packages: locket, pluggy, partd, fsspec, zstandard, pytest, Deprecated, pandas-plink, genoml2\n",
            "  Attempting uninstall: pluggy\n",
            "    Found existing installation: pluggy 0.7.1\n",
            "    Uninstalling pluggy-0.7.1:\n",
            "      Successfully uninstalled pluggy-0.7.1\n",
            "  Attempting uninstall: pytest\n",
            "    Found existing installation: pytest 3.6.4\n",
            "    Uninstalling pytest-3.6.4:\n",
            "      Successfully uninstalled pytest-3.6.4\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.\u001b[0m\n",
            "Successfully installed Deprecated-1.2.13 fsspec-2022.2.0 genoml2-1.0.0b11 locket-0.2.1 pandas-plink-2.2.9 partd-1.2.0 pluggy-1.0.0 pytest-7.1.0 zstandard-0.17.0\n",
            "Requirement already satisfied: tables in /usr/local/lib/python3.7/dist-packages (3.7.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.7/dist-packages (from tables) (21.3)\n",
            "Requirement already satisfied: numpy>=1.19.0 in /usr/local/lib/python3.7/dist-packages (from tables) (1.21.5)\n",
            "Requirement already satisfied: numexpr>=2.6.2 in /usr/local/lib/python3.7/dist-packages (from tables) (2.8.1)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging->tables) (3.0.7)\n",
            "Collecting umap-learn\n",
            "  Downloading umap-learn-0.5.2.tar.gz (86 kB)\n",
            "\u001b[K     |████████████████████████████████| 86 kB 3.9 MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy>=1.17 in /usr/local/lib/python3.7/dist-packages (from umap-learn) (1.21.5)\n",
            "Requirement already satisfied: scikit-learn>=0.22 in /usr/local/lib/python3.7/dist-packages (from umap-learn) (1.0.2)\n",
            "Requirement already satisfied: scipy>=1.0 in /usr/local/lib/python3.7/dist-packages (from umap-learn) (1.4.1)\n",
            "Requirement already satisfied: numba>=0.49 in /usr/local/lib/python3.7/dist-packages (from umap-learn) (0.51.2)\n",
            "Collecting pynndescent>=0.5\n",
            "  Downloading pynndescent-0.5.6.tar.gz (1.1 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.1 MB 18.4 MB/s \n",
            "\u001b[?25hRequirement already satisfied: tqdm in /usr/local/lib/python3.7/dist-packages (from umap-learn) (4.63.0)\n",
            "Requirement already satisfied: llvmlite<0.35,>=0.34.0.dev0 in /usr/local/lib/python3.7/dist-packages (from numba>=0.49->umap-learn) (0.34.0)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from numba>=0.49->umap-learn) (57.4.0)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.7/dist-packages (from pynndescent>=0.5->umap-learn) (1.1.0)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.22->umap-learn) (3.1.0)\n",
            "Building wheels for collected packages: umap-learn, pynndescent\n",
            "  Building wheel for umap-learn (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for umap-learn: filename=umap_learn-0.5.2-py3-none-any.whl size=82708 sha256=12328b52b1ffba01ee4082b8e7a8370fa1d00dfc5102c4cfb389658634792eb1\n",
            "  Stored in directory: /root/.cache/pip/wheels/84/1b/c6/aaf68a748122632967cef4dffef68224eb16798b6793257d82\n",
            "  Building wheel for pynndescent (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pynndescent: filename=pynndescent-0.5.6-py3-none-any.whl size=53943 sha256=69b16165d14bb53b96f17597aae607c26253733af0de7ec05bc1d84cec904913\n",
            "  Stored in directory: /root/.cache/pip/wheels/03/f1/56/f80d72741e400345b5a5b50ec3d929aca581bf45e0225d5c50\n",
            "Successfully built umap-learn pynndescent\n",
            "Installing collected packages: pynndescent, umap-learn\n",
            "Successfully installed pynndescent-0.5.6 umap-learn-0.5.2\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eASrIwvFbzju",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5c81ba23-eb47-4e02-a8ea-f69a4b9d33bd"
      },
      "source": [
        "# Munging\n",
        "\n",
        "%%bash\n",
        "genoml discrete supervised munge \\\n",
        "--prefix outputs/test_discrete_geno \\\n",
        "--geno ./discrete/validation \\\n",
        "--pheno ./discrete/validation_pheno.csv"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b6.16 64-bit (19 Feb 2020)          www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to plink.log.\n",
            "Options in effect:\n",
            "  --bfile ./discrete/validation\n",
            "  --indep-pairwise 1000 50 0.5\n",
            "\n",
            "12991 MB RAM detected; reserving 6495 MB for main workspace.\n",
            "500 variants loaded from .bim file.\n",
            "100 people (67 males, 33 females) loaded from .fam.\n",
            "100 phenotype values loaded from .fam.\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 100 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "500 variants and 100 people pass filters and QC.\n",
            "Among remaining phenotypes, 73 are cases and 27 are controls.\n",
            "\rPruned 0 variants from chromosome 1, leaving 48.\n",
            "\rPruned 1 variant from chromosome 2, leaving 40.\n",
            "\rPruned 1 variant from chromosome 3, leaving 45.\n",
            "\rPruned 1 variant from chromosome 4, leaving 39.\n",
            "\rPruned 0 variants from chromosome 5, leaving 30.\n",
            "\rPruned 0 variants from chromosome 6, leaving 40.\n",
            "\rPruned 0 variants from chromosome 7, leaving 14.\n",
            "\rPruned 0 variants from chromosome 8, leaving 14.\n",
            "\rPruned 0 variants from chromosome 9, leaving 17.\n",
            "\rPruned 2 variants from chromosome 10, leaving 27.\n",
            "\rPruned 3 variants from chromosome 11, leaving 27.\n",
            "\rPruned 0 variants from chromosome 12, leaving 24.\n",
            "\rPruned 0 variants from chromosome 13, leaving 9.\n",
            "\rPruned 2 variants from chromosome 14, leaving 17.\n",
            "\rPruned 0 variants from chromosome 15, leaving 6.\n",
            "\rPruned 1 variant from chromosome 16, leaving 15.\n",
            "\rPruned 0 variants from chromosome 17, leaving 26.\n",
            "\rPruned 0 variants from chromosome 18, leaving 9.\n",
            "\rPruned 0 variants from chromosome 19, leaving 12.\n",
            "\rPruned 0 variants from chromosome 20, leaving 11.\n",
            "\rPruned 0 variants from chromosome 21, leaving 6.\n",
            "\rPruned 1 variant from chromosome 22, leaving 12.\n",
            "Pruning complete.  12 of 500 variants removed.\n",
            "Writing...\rMarker lists written to plink.prune.in and plink.prune.out .\n",
            "PLINK v1.90b6.16 64-bit (19 Feb 2020)          www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to temp_genos.log.\n",
            "Options in effect:\n",
            "  --bfile ./discrete/validation\n",
            "  --extract plink.prune.in\n",
            "  --make-bed\n",
            "  --out temp_genos\n",
            "\n",
            "12991 MB RAM detected; reserving 6495 MB for main workspace.\n",
            "500 variants loaded from .bim file.\n",
            "100 people (67 males, 33 females) loaded from .fam.\n",
            "100 phenotype values loaded from .fam.\n",
            "--extract: 488 variants remaining.\n",
            "Using 1 thread (no multithreaded calculations invoked).\n",
            "Before main variant filters, 100 founders and 0 nonfounders present.\n",
            "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\n",
            "488 variants and 100 people pass filters and QC.\n",
            "Among remaining phenotypes, 73 are cases and 27 are controls.\n",
            "--make-bed to temp_genos.bed + temp_genos.bim + temp_genos.fam ... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\n",
            "\n",
            "\u001b[34m---> \u001b[0m\u001b[34mRunning genoml discrete supervised munge\u001b[0m\n",
            "\n",
            "\u001b[34m---> ---> \u001b[0m\u001b[34mDependency Check\u001b[0m\n",
            "\n",
            "\u001b[34m---> ---> ---> \u001b[0m\u001b[34mChecking plink\u001b[0m\n",
            "\u001b[32m---> ---> \u001b[0m\u001b[32mDependency Check\u001b[0m: \u001b[32m[Done]\u001b[0m\n",
            "\n",
            "\u001b[34m---> ---> \u001b[0m\u001b[34mDependency Check\u001b[0m\n",
            "\n",
            "\u001b[34m---> ---> ---> \u001b[0m\u001b[34mChecking plink\u001b[0m\n",
            "\u001b[32m---> ---> \u001b[0m\u001b[32mDependency Check\u001b[0m: \u001b[32m[Done]\u001b[0m\n",
            "\n",
            "Here is some basic info on the command you are about to run.\n",
            "Python version info...\n",
            "3.7.12 (default, Jan 15 2022, 18:48:18) \n",
            "[GCC 7.5.0]\n",
            "CLI argument info...\n",
            "The output prefix for this run is outputs/test_discrete_geno and will be appended to later runs of GenoML.\n",
            "Working with genotype data? ./discrete/validation\n",
            "Do you want GenoML to prune your SNPs for you? no\n",
            "The pruning threshold you've chosen is 0.5\n",
            "Working with additional predictors? None\n",
            "Where is your phenotype file? ./discrete/validation_pheno.csv\n",
            "Any use for an external set of GWAS summary stats? None\n",
            "If you plan on using external GWAS summary stats for SNP filtering, we'll only keep SNPs at what P value? 0.001\n",
            "How strong is your VIF filter? 0\n",
            "How many iterations of VIF filtering are you doing? 0\n",
            "The imputation method you picked is using the column median to fill in any remaining NAs.\n",
            "Will you be adjusting additional features using UMAP dimensionality reduction? no\n",
            "Give credit where credit is due, for this stage of analysis we use code from the great contributors to python packages: os, sys, argparse, numpy, pandas, joblib, math and time. We also use PLINK v1.9 from https://www.cog-genomics.org/plink/1.9/.\n",
            "\n",
            "No additional features as predictors? No problem, we'll stick to genotypes.\n",
            "So you don't want to filter on P values from external GWAS? No worries, we don't usually either (if the dataset is large enough).\n",
            "Exporting genotype data\n",
            "A list of pruned variants and the allele being counted in the dosages (usually the minor allele) can be found here: outputs/test_discrete_geno.variants_and_alleles.tab\n",
            "rm temp_genos.*\n",
            "\n",
            "You have just imputed your genotype features, covering up NAs with the column median so that analyses don't crash due to missing data.\n",
            "Now your genotype features might look a little better (showing the first few lines of the left-most and right-most columns)...\n",
            "######################################################################\n",
            "variant      snp410      snp403      snp164  ...      snp368      snp304      snp306\n",
            "count    100.000000  100.000000  100.000000  ...  100.000000  100.000000  100.000000\n",
            "mean       0.830000    0.100000    0.920000  ...    0.340000    0.300000    0.040000\n",
            "std        0.652191    0.301511    0.691653  ...    0.535978    0.541229    0.196946\n",
            "min        0.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000\n",
            "25%        0.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000\n",
            "50%        1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000\n",
            "75%        1.000000    0.000000    1.000000  ...    1.000000    1.000000    0.000000\n",
            "max        2.000000    1.000000    2.000000  ...    2.000000    2.000000    1.000000\n",
            "\n",
            "[8 rows x 488 columns]\n",
            "######################################################################\n",
            "\n",
            "An updated list of 490 features, including ID and PHENO, that is in your munged dataForML.h5 file can be found here outputs/test_discrete_geno.list_features.txt\n",
            "\n",
            "Your .dataForML file that has been fully munged can be found here: outputs/test_discrete_geno.dataForML.h5\n",
            "Thank you for munging with GenoML!\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
            "  import pandas.util.testing as tm\n",
            "WARNING:root:Installing Plink\n",
            "\rMapping files:   0%|          | 0/3 [00:00<?, ?it/s]\rMapping files: 100%|██████████| 3/3 [00:00<00:00, 149.82it/s]\n",
            "/usr/local/lib/python3.7/dist-packages/genoml/preprocessing/munging.py:213: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.\n",
            "  raw_df = raw_df.fillna(raw_df.median())\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IJJZg94FeHBA",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "2392bf61-5807-4f58-ec1a-4ce4b3939d09"
      },
      "source": [
        "# Training\n",
        "%%bash\n",
        "genoml discrete supervised train \\\n",
        "--prefix outputs/test_discrete_geno #This doesn't seem to work"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\u001b[34m---> \u001b[0m\u001b[34mRunning genoml discrete supervised train\u001b[0m\n",
            "\n",
            "\u001b[34m---> ---> \u001b[0m\u001b[34mDependency Check\u001b[0m\n",
            "\n",
            "\u001b[34m---> ---> ---> \u001b[0m\u001b[34mChecking plink\u001b[0m\n",
            "\u001b[32m---> ---> \u001b[0m\u001b[32mDependency Check\u001b[0m: \u001b[32m[Done]\u001b[0m\n",
            "\n",
            "Here is some basic info on the command you are about to run.\n",
            "Python Version info...\n",
            "3.7.12 (default, Jan 15 2022, 18:48:18) \n",
            "[GCC 7.5.0]\n",
            "CLI argument info...\n",
            "Working with dataset outputs/test_discrete_geno from previous data munging efforts.\n",
            "You have chosen to compete the algorithms based on AUC.\n",
            "Give credit where credit is due, for this stage of analysis we use code from the great contributors to Python packages: argparse, xgboost, sklearn, pandas, numpy, time, matplotlib and seaborn.\n",
            "As a note, in all exported probabilities and other graphics, case status is treated as a 0 or 1, with 1 representing a positive case.\n",
            "\n",
            "Your data looks like this (showing the first few lines of the left-most and right-most columns)...\n",
            "######################################################################\n",
            "            PHENO      snp410      snp403  ...      snp368      snp304      snp306\n",
            "count  100.000000  100.000000  100.000000  ...  100.000000  100.000000  100.000000\n",
            "mean     0.730000    0.830000    0.100000  ...    0.340000    0.300000    0.040000\n",
            "std      0.446196    0.652191    0.301511  ...    0.535978    0.541229    0.196946\n",
            "min      0.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000\n",
            "25%      0.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000\n",
            "50%      1.000000    1.000000    0.000000  ...    0.000000    0.000000    0.000000\n",
            "75%      1.000000    1.000000    0.000000  ...    1.000000    1.000000    0.000000\n",
            "max      1.000000    2.000000    1.000000  ...    2.000000    2.000000    1.000000\n",
            "\n",
            "[8 rows x 489 columns]\n",
            "######################################################################\n",
            "\n",
            "Now let's compete these algorithms!\n",
            "We'll update you as each algorithm runs, then summarize at the end.\n",
            "Here we test each algorithm under default settings using the same training and test datasets derived from a 70% training and 30% testing split of your data.\n",
            "For each algorithm, we will output the following metrics...\n",
            "Algorithm name, hoping that's pretty self-explanatory. Plenty of resources on these common ML algorithms at https://scikit-learn.org and https://xgboost.readthedocs.io/.\n",
            "AUC_percent, this is the area under the curve from receiver operating characteristic analyses. This is the most common metric of classifier performance in biomedical literature, we express this as a percent. We calculate AUC based on the predicted probability of being a case.\n",
            "Accuracy_percent, this is the simple accuracy of the classifier, how many predictions were correct from best classification cutoff (python default).\n",
            "Balanced_Accuracy_Percent, consider this as the accuracy resampled to a 1:1 mix of cases and controls. Imbalanced datasets can give funny results for simple accuracy.\n",
            "Log_Loss, this is essentially the inverse of the likelihood function for a correct prediction, you want to minimize this.\n",
            "Sensitivity, proportion of cases correctly identified.\n",
            "Specificity, proportion of controls correctly identified.\n",
            "PPV, this is the positive predictive value, the probability that subjects with a positive result actually have the disease.\n",
            "NPV, this is the negative predictive value, the probability that subjects with a negative result don't have the disease.\n",
            "We also log the runtimes per algorithm.\n",
            "\n",
            "Algorithm summaries incoming...\n",
            "\n",
            "\n",
            "######################################################################\n",
            "\n",
            "LogisticRegression\n",
            "AUC: 45.4545%\n",
            "Accuracy: 63.3333%\n",
            "Balanced Accuracy: 47.1591%\n",
            "Log Loss: 0.8876\n",
            "Runtime in seconds: 0.1147\n",
            "\n",
            "######################################################################\n",
            "\n",
            "RandomForestClassifier\n",
            "AUC: 53.1250%\n",
            "Accuracy: 73.3333%\n",
            "Balanced Accuracy: 50.0000%\n",
            "Log Loss: 0.5734\n",
            "Runtime in seconds: 0.4042\n",
            "\n",
            "######################################################################\n",
            "\n",
            "AdaBoostClassifier\n",
            "AUC: 48.8636%\n",
            "Accuracy: 56.6667%\n",
            "Balanced Accuracy: 38.6364%\n",
            "Log Loss: 0.6268\n",
            "Runtime in seconds: 0.1383\n",
            "\n",
            "######################################################################\n",
            "\n",
            "GradientBoostingClassifier\n",
            "AUC: 59.0909%\n",
            "Accuracy: 70.0000%\n",
            "Balanced Accuracy: 59.6591%\n",
            "Log Loss: 1.254\n",
            "Runtime in seconds: 0.2191\n",
            "\n",
            "######################################################################\n",
            "\n",
            "SGDClassifier\n",
            "AUC: 42.6136%\n",
            "Accuracy: 56.6667%\n",
            "Balanced Accuracy: 42.6136%\n",
            "Log Loss: 14.97\n",
            "Runtime in seconds: 0.03267\n",
            "\n",
            "######################################################################\n",
            "\n",
            "SVC\n",
            "AUC: 56.2500%\n",
            "Accuracy: 73.3333%\n",
            "Balanced Accuracy: 50.0000%\n",
            "Log Loss: 0.5808\n",
            "Runtime in seconds: 0.04818\n",
            "\n",
            "######################################################################\n",
            "\n",
            "MLPClassifier\n",
            "AUC: 39.7727%\n",
            "Accuracy: 70.0000%\n",
            "Balanced Accuracy: 47.7273%\n",
            "Log Loss: 1.169\n",
            "Runtime in seconds: 0.2693\n",
            "\n",
            "######################################################################\n",
            "\n",
            "KNeighborsClassifier\n",
            "AUC: 62.5000%\n",
            "Accuracy: 60.0000%\n",
            "Balanced Accuracy: 56.8182%\n",
            "Log Loss: 0.6071\n",
            "Runtime in seconds: 0.04177\n",
            "\n",
            "######################################################################\n",
            "\n",
            "LinearDiscriminantAnalysis\n",
            "AUC: 47.1591%\n",
            "Accuracy: 73.3333%\n",
            "Balanced Accuracy: 50.0000%\n",
            "Log Loss: 0.5994\n",
            "Runtime in seconds: 0.1069\n",
            "\n",
            "######################################################################\n",
            "\n",
            "QuadraticDiscriminantAnalysis\n",
            "AUC: 43.7500%\n",
            "Accuracy: 46.6667%\n",
            "Balanced Accuracy: 43.7500%\n",
            "Log Loss: 18.42\n",
            "Runtime in seconds: 0.08161\n",
            "\n",
            "######################################################################\n",
            "\n",
            "BaggingClassifier\n",
            "AUC: 25.0000%\n",
            "Accuracy: 73.3333%\n",
            "Balanced Accuracy: 53.9773%\n",
            "Log Loss: 2.9\n",
            "Runtime in seconds: 0.08851\n",
            "\n",
            "######################################################################\n",
            "\n",
            "XGBClassifier\n",
            "AUC: 49.4318%\n",
            "Accuracy: 63.3333%\n",
            "Balanced Accuracy: 43.1818%\n",
            "Log Loss: 0.7947\n",
            "Runtime in seconds: 0.2612\n",
            "######################################################################\n",
            "\n",
            "...remember, there are occasionally slight fluctuations in model performance on the same withheld samples...\n",
            "######################################################################\n",
            "KNeighborsClassifier\n",
            "AUC: 49.4318%\n",
            "Accuracy: 60.0000%\n",
            "Balanced Accuracy: 56.8182%\n",
            "Log Loss: 0.7947\n",
            "######################################################################\n",
            "... this model has been saved as outputs/test_discrete_geno.trainedModel.joblib for later use and can be found in your working directory.\n",
            "We are also exporting a ROC curve for you here outputs/test_discrete_geno.trainedModel_withheldSample_ROC.png this is a graphical representation of AUC in the withheld test data for the best performing algorithm.\n",
            "\n",
            "Preview of the exported predictions for the withheld test data that has been exported as outputs/test_discrete_geno.trainedModel_withheldSample_Predictions.csv these are pretty straight forward.\n",
            "They generally include the sample ID, the previously reported case status (1 = case), the case probability from the best performing algorithm and the predicted label from that algorithm\n",
            "\n",
            "######################################################################\n",
            "             ID  CASE_REPORTED  CASE_PROBABILITY  CASE_PREDICTED\n",
            "0  valdiation84              1               0.6               1\n",
            "1  valdiation54              0               0.4               0\n",
            "2  valdiation71              1               0.4               0\n",
            "3  valdiation46              1               0.6               1\n",
            "4  valdiation45              1               0.4               0\n",
            "######################################################################\n",
            "\n",
            "Preview of the exported predictions for the training samples which is naturally overfit and exported as outputs/test_discrete_geno.trainedModel_trainingSample_Predictions.csv in the similar format as in the withheld test dataset that was just exported.\n",
            "######################################################################\n",
            "             ID  CASE_REPORTED  CASE_PROBABILITY  CASE_PREDICTED\n",
            "0  valdiation12              1               0.6               1\n",
            "1  valdiation48              0               0.4               0\n",
            "2  valdiation86              0               0.6               1\n",
            "3  valdiation29              0               0.6               1\n",
            "4  valdiation94              1               0.8               1\n",
            "######################################################################\n",
            "We are also exporting probability density plots to the file outputs/test_discrete_geno.trainedModel_withheldSample_probabilities.png this is a plot of the probability distributions of being a case, stratified by case and control status in the withheld test samples.\n",
            "A complete table of the performance metrics can be found at outputs/test_discrete_geno.training_withheldSamples_performanceMetrics.csv\n",
            "            Note that any models that were overfit (if AUC or Balanced Accuracy was 50% or less, or sensitivity/specificity were 1 or 0) were not considered when nominating the best algorithm.\n",
            "This table below is also logged as outputs/test_discrete_geno.training_withheldSamples_performanceMetrics.csv and is in your current working directory...\n",
            "######################################################################\n",
            "                       Algorithm  AUC_Percent  ...       NPV  Runtime_Seconds\n",
            "0             LogisticRegression    45.454545  ...  0.200000         0.114659\n",
            "0         RandomForestClassifier    53.125000  ...       NaN         0.404204\n",
            "0             AdaBoostClassifier    48.863636  ...  0.000000         0.138298\n",
            "0     GradientBoostingClassifier    59.090909  ...  0.428571         0.219065\n",
            "0                  SGDClassifier    42.613636  ...  0.142857         0.032674\n",
            "0                            SVC    56.250000  ...       NaN         0.048184\n",
            "0                  MLPClassifier    39.772727  ...  0.000000         0.269344\n",
            "0           KNeighborsClassifier    62.500000  ...  0.333333         0.041774\n",
            "0     LinearDiscriminantAnalysis    47.159091  ...       NaN         0.106909\n",
            "0  QuadraticDiscriminantAnalysis    43.750000  ...  0.214286         0.081611\n",
            "0              BaggingClassifier    25.000000  ...  0.500000         0.088513\n",
            "0                  XGBClassifier    49.431818  ...  0.000000         0.261229\n",
            "\n",
            "[12 rows x 10 columns]\n",
            "######################################################################\n",
            "Based on your withheld samples, the algorithm with the best AUC is the KNeighborsClassifier... let's save that model for you.\n",
            "Thank you for training with GenoML!\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
            "  import pandas.util.testing as tm\n",
            "/usr/local/lib/python3.7/dist-packages/genoml/discrete/supervised/training.py:117: RuntimeWarning: invalid value encountered in long_scalars\n",
            "  NPV = TN/(TN+FN)\n",
            "/usr/local/lib/python3.7/dist-packages/genoml/discrete/supervised/training.py:117: RuntimeWarning: invalid value encountered in long_scalars\n",
            "  NPV = TN/(TN+FN)\n",
            "/usr/local/lib/python3.7/dist-packages/genoml/discrete/supervised/training.py:117: RuntimeWarning: invalid value encountered in long_scalars\n",
            "  NPV = TN/(TN+FN)\n",
            "/usr/local/lib/python3.7/dist-packages/sklearn/discriminant_analysis.py:878: UserWarning: Variables are collinear\n",
            "  warnings.warn(\"Variables are collinear\")\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-tX3tmYVeLR0",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b539845d-3f72-4b4c-c2b5-2749b21fedc9"
      },
      "source": [
        "# Tuning\n",
        "%%bash\n",
        "genoml discrete supervised tune \\\n",
        "--prefix outputs/test_discrete_geno"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\u001b[34m---> \u001b[0m\u001b[34mRunning genoml discrete supervised tune\u001b[0m\n",
            "\n",
            "\u001b[34m---> ---> \u001b[0m\u001b[34mDependency Check\u001b[0m\n",
            "\n",
            "\u001b[34m---> ---> ---> \u001b[0m\u001b[34mChecking plink\u001b[0m\n",
            "\u001b[32m---> ---> \u001b[0m\u001b[32mDependency Check\u001b[0m: \u001b[32m[Done]\u001b[0m\n",
            "Here is some basic info on the command you are about to run.\n",
            "CLI argument info...\n",
            "Working with the dataset and best model corresponding to prefix outputs/test_discrete_geno the timestamp from the merge is the prefix in most cases.\n",
            "You have chosen to tune the algorithms based on AUC.\n",
            "Your maximum number of tuning iterations is 50 and if you are concerned about runtime, make this number smaller.\n",
            "You are running 5 rounds of cross-validation, and again... if you are concerned about runtime, make this number smaller.\n",
            "Give credit where credit is due, for this stage of analysis we use code from the great contributors to python packages: argparse, xgboost, sklearn, pandas, numpy, time, matplotlib and seaborn.\n",
            "As a note, in all exported probabilities and other graphics, case status is treated as a 0 or 1, with 1 representing a positive case.\n",
            "\n",
            "From previous analyses in the training phase, we've determined that the best algorithm for this application is KNeighborsClassifier... so let's tune it up and see what gains we can make!\n",
            "Here is a summary of the top 10 iterations of the hyperparameter tuning...\n",
            "RandomizedSearchCV took 4.34 seconds for 50 candidates parameter iterations.\n",
            "Model with Rank: 1\n",
            "Mean Validation Score: 0.677 (std: 0.098)\n",
            "Parameters: {'leaf_size': 45, 'n_neighbors': 6}\n",
            "\n",
            "Model with Rank: 1\n",
            "Mean Validation Score: 0.677 (std: 0.098)\n",
            "Parameters: {'leaf_size': 29, 'n_neighbors': 6}\n",
            "\n",
            "Model with Rank: 1\n",
            "Mean Validation Score: 0.677 (std: 0.098)\n",
            "Parameters: {'leaf_size': 8, 'n_neighbors': 6}\n",
            "\n",
            "Model with Rank: 1\n",
            "Mean Validation Score: 0.677 (std: 0.098)\n",
            "Parameters: {'leaf_size': 10, 'n_neighbors': 6}\n",
            "\n",
            "Model with Rank: 1\n",
            "Mean Validation Score: 0.677 (std: 0.098)\n",
            "Parameters: {'leaf_size': 3, 'n_neighbors': 6}\n",
            "\n",
            "Model with Rank: 6\n",
            "Mean Validation Score: 0.651 (std: 0.102)\n",
            "Parameters: {'leaf_size': 99, 'n_neighbors': 5}\n",
            "\n",
            "Model with Rank: 6\n",
            "Mean Validation Score: 0.651 (std: 0.102)\n",
            "Parameters: {'leaf_size': 58, 'n_neighbors': 5}\n",
            "\n",
            "Model with Rank: 6\n",
            "Mean Validation Score: 0.651 (std: 0.102)\n",
            "Parameters: {'leaf_size': 77, 'n_neighbors': 5}\n",
            "\n",
            "Model with Rank: 6\n",
            "Mean Validation Score: 0.651 (std: 0.102)\n",
            "Parameters: {'leaf_size': 34, 'n_neighbors': 5}\n",
            "\n",
            "Model with Rank: 6\n",
            "Mean Validation Score: 0.651 (std: 0.102)\n",
            "Parameters: {'leaf_size': 3, 'n_neighbors': 5}\n",
            "\n",
            "Model with Rank: 6\n",
            "Mean Validation Score: 0.651 (std: 0.102)\n",
            "Parameters: {'leaf_size': 62, 'n_neighbors': 5}\n",
            "\n",
            "Model with Rank: 6\n",
            "Mean Validation Score: 0.651 (std: 0.102)\n",
            "Parameters: {'leaf_size': 65, 'n_neighbors': 5}\n",
            "\n",
            "We are exporting a summary table of the top 10 iterations of the hyperparameter tuning step and its parameters here outputs/test_discrete_geno.tunedModel_top10Iterations_Summary.csv.\n",
            "Here is the cross-validation summary of your best tuned model hyperparameters...\n",
            "Scores per cross-validation of the metric to be maximized, this scoring metric is AUC or Balanced_Accuracy for discrete phenotypes and explained variance for continuous phenotypes:\n",
            "[0.60119048 0.83928571 0.72       0.66666667 0.56      ]\n",
            "Mean cross-validation score:\n",
            "0.6774285714285715\n",
            "Standard deviation of the cross-validation score:\n",
            "0.09770117335209651\n",
            "\n",
            "Here is the cross-validation summary of your baseline/default hyperparamters for the same algorithm on the same data...\n",
            "Scores per cross-validation of the metric to be maximized, this scoring metric is AUC or Balanced_Accuracy for discrete phenotypes and explained variance for continuous phenotypes:\n",
            "[0.61309524 0.58928571 0.82       0.70666667 0.52666667]\n",
            "Mean cross-validation score:\n",
            "0.651142857142857\n",
            "Standard deviation of the cross-validation score:\n",
            "0.10233657953148563\n",
            "\n",
            "Just a note, if you have a relatively small variance among the cross-validation iterations, there is a higher chance of your model being more generalizable to similar datasets.\n",
            "We are exporting a summary table of the cross-validation mean score and standard deviation of the baseline vs. best tuned model here outputs/test_discrete_geno.tunedModel_CV_Summary.csv.\n",
            "\n",
            "Based on comparisons of the default parameters to your hyperparameter tuned model, the tuned model actually performed better.\n",
            "Looks like the tune was worth it, we suggest using this model for maximum performance, lets summarize and export this now.\n",
            "In most cases, if opting to use the tuned model, a separate test dataset is a good idea. GenoML has a module to fit models to external data.\n",
            "\n",
            "Preview of the exported predictions for the tuning samples which is naturally overfit and exported as outputs/test_discrete_geno.tunedModel_allSample_Predictions.csv in the similar format as in the initial training phase of GenoML.\n",
            "######################################################################\n",
            "            ID  CASE_REPORTED  CASE_PROBABILITY  CASE_PREDICTED\n",
            "0  valdiation1              0          0.833333               1\n",
            "1  valdiation2              1          0.500000               0\n",
            "2  valdiation3              1          0.500000               0\n",
            "3  valdiation4              1          0.833333               1\n",
            "4  valdiation5              1          1.000000               1\n",
            "######################################################################\n",
            "We are also exporting probability density plots to the file outputs/test_discrete_geno.tunedModel_allSample_probabilities.png this is a plot of the probability distributions of being a case, stratified by case and control status for all samples.\n",
            "\n",
            "End of tuning stage with GenoML.\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
            "  import pandas.util.testing as tm\n"
          ]
        }
      ]
    }
  ]
}